text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+typecheck time. By limiting which values a trait is implemented over, the
+orphan rules can enable a crate to ensure that only some safe values are used,
+with the check performed at compile time (this is especially relevant to
+cryptographic libraries for example).
+
+# Detailed design
+[design]: #detailed-design
+
+Today, types in Rust can be parameterized by two kinds: types and lifetimes. We
+will additionally allow types to be parameterized by values, so long as those
+values can be computed at compile time. A single constant parameter must be of
+a single, particular type, and can be validly substituted with any value of
+that type which can be computed at compile time and the type meets the equality
+requirements laid out later in this RFC.
+
+(Exactly which expressions are evaluable at compile time is orthogonal to this
+RFC. For our purposes we assume that integers and their basic arithmetic
+operations can be computed at compile time, and we will use them in all
+examples.)
+
+## Glossary
+
+* __Const (constant, const value):__ A Rust value which is guaranteed to be
+fully evaluated at compile time. Unlike statics, consts will be inlined at
+their use sites rather than existing in the data section of the compiled
+binary.
+
+* __Const parameter (generic const):__ A const which a type or function is
+abstract over; this const is input to the concrete type of the item, such as
+the length parameter of a static array.
+
+* __Associated const:__ A const associated with a trait, similar to an
+associated type. Unlike a const parameter, an associated const is *determined*
+by a type.
+
+* __Const variable:__ Either a const parameter or a trait, contrast with
+concrete const; a const which is undetermined in this context (prior to
+monomorphization).
+
+* __Concrete const:__ In contrast to a const variable, a const which has a
+known and singular value in this context.
+
+* __Const expression:__ An expression which evaluates to a const. This may be
+an identity expression or a more complex expression, so long as it can be
+evaluated by Rust's const system.
+
+* __Abstract const expression:__ A const expression which involves a const
+variable (and therefore the value that it evaluates to cannot be determined
+until after monomorphization).
+
+* __Const projection:__ The value of an abstract const expression (which cannot
+be determined in a generic context because it is dependent on a const
+variable).
+
+* __Identity expression:__ An expression which cannot be evaluated further
+except by substituting it with names in scope. This includes all literals as
+well all idents - e.g. `3`, `"Hello, world"`, `foo_bar`.
+
+## Declaring a const parameter
+
+In any sequence of type parameter declarations (such as in the definition of a
+type or on the `impl` header of an impl block) const parameters can also be
+declared. Const parameters always come after type parameters, and their
```

Comment:
I don't feel like I've ever properly understood this concern. Can't we determine from the kind of the node you put there and (usually) how many arguments you've supplied whether it is intended to be a const or a type?

The only case that seems ambiguous to me is something like this:

```rust
struct Foo<T = i32, const N: usize = 0>([T; N]);

fn foo<T, const T: usize>(_: Foo<T>) { }
```

That is you have both const and type default parameters, and you have an ident which is a name in both type and const context (bad news in general), and you supply it once to the type. I don't particularly care what we do here since its such an edge case (probably treat it as the type parameter).

Am I missing something? Why wouldn't this Just Work?